src/publish.ts

```diff
@@ -0,0 +1,323 @@
+import * as core from '@actions/core'
+import * as exec from '@actions/exec'
+import * as yaml from 'js-yaml'
+
+import * as github from '@actions/github'
+import { Plan } from './model'
+import { DefaultArtifactClient } from '@actions/artifact'
+import fs from 'fs'
+import path from 'path'
+import os from 'os'
+
+class Publish {
+  private token: string
+  private charmhubToken: string
+  private octokit
+  private workflowFile: string
+  private artifact
+  private workingDir: string
+  private resourceMapping: { [key: string]: string }
+  private workflowRunId: string
+  private identifier: string
+
+  constructor() {
+    this.token = core.getInput('github-token')
+    this.charmhubToken = core.getInput('charmhub-token')
+    this.octokit = github.getOctokit(this.token)
+    this.workflowFile = core.getInput('workflow-file')
+    this.workingDir = core.getInput('working-directory')
+    this.resourceMapping = JSON.parse(core.getInput('resource-mapping'))
+    this.workflowRunId = core.getInput('workflow-run-id')
+    this.identifier = core.getInput('identifier')
+    this.artifact = new DefaultArtifactClient()
+  }
+
+  async findWorkflowRunId(): Promise<number> {
```

Comment:
You may want to use step-down rules and ordering features to improve readability: https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#function-and-method-ordering